### PR TITLE
Ksql 669 Defer topic existence validation to the physical plan generation

### PR DESCRIPTION
**Pull request by hjafarpour**
This will fix the issue #669  where we a new stream/table DDL statement is defined on the output of another CSAS/CTAS statement. We will only enforce existence of a topic in a CREATE STREAM/TABLE statement when we actually create the stream/table and not during parsing and validation of the statements in the script.

### CODE REVIEW

**Comment by dguy** [comment]
tests?

**Comment by hjafarpour** [comment]
@dguy @apurvam @rodesai @bluemonk3y friendly reminder for review! :)

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java [hunk 1]
```diff
@@ -621,4 +624,12 @@ private void executeDDLCommand(DDLCommand ddlCommand) {
         );
     return avroCheckResult;
   }
+
+  private KsqlEntity handleRunScript(Map<String, Object> streamsProperties) {
```
Comment (rodesai) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java]: I don't see this being called anywhere. And it doesn't seem to actually be doing anything. Leftover?

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 2]
```diff
@@ -275,12 +278,13 @@ public QueryMetadata getQueryExecutionPlan(final Query query) throws Exception {
     }
   }
 
-  private Pair<String, Statement> buildSingleQueryAst(
+  public Pair<String, Statement> buildSingleQueryAst(
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: Why this change?

File: ksql-common/src/main/java/io/confluent/ksql/ddl/DdlConfig.java [hunk 3]
```diff
@@ -29,4 +29,6 @@
   public static final String TIMESTAMP_NAME_PROPERTY = "TIMESTAMP";
   public static final String PARTITION_BY_PROPERTY = "PARTITION_BY";
   public static final String SCHEMA_FILE_CONTENT_PROPERTY = "ksql.schema.file.content";
+  public static final String RUN_SCRIPT_STATEMENTS_CONTENT = "ksql.run.script.statements";
+  public static final String ENFORCE_TOPIC_EXISTANCE = "ENFORCE_TOPIC_EXISTANCE";
```
Comment (dguy) [inline ksql-common/src/main/java/io/confluent/ksql/ddl/DdlConfig.java]: This doesn't look like it used

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java [hunk 4]
```diff
@@ -169,7 +172,8 @@ protected void checkMetaData(MetaStore metaStore, String sourceName, String topi
 
   protected RegisterTopicCommand registerTopicFirst(
       Map<String, Expression> properties,
-      Map<String, Object> overriddenProperties
+      Map<String, Object> overriddenProperties,
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java]: this is not used

File: ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java [hunk 5]
```diff
@@ -74,4 +77,19 @@ public void shouldFailWhenSyntaxIsInvalid() throws Exception {
...
 
+  @Test
+  public void shouldEnforceTopicExistanceCorrectly() throws Exception {
+    topicClient.createTopic("s1_topic", 1, (short) 1);
+    StringBuilder runScriptContent =
+        new StringBuilder("CREATE STREAM S1 (COL1 BIGINT, COL2 VARCHAR) "
+                          + "WITH  (KAFKA_TOPIC = 's1_topic', VALUE_FORMAT = 'JSON');\n");
+    runScriptContent.append("CREATE TABLE T1 AS SELECT COL1, count(*) FROM "
+                            + "S1 GROUP BY COL1;\n");
+    runScriptContent.append("CREATE STREAM S2 (C1 BIGINT, C2 BIGINT) "
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java]: We probably also want to add a test to check that we fail in the case where the topic really doesn't exist. I can see that we drop the topic existence check during parsing. But I am not sure where we actually add that check back. 

In particular, if we have a `CREATE STREAM` DDL statement without an underlying topic, will we fail with this patch? I am not sure anymore

Comment (apurvam) [reply ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java]: I guess the check for existence happens when we actually execute a DDL statement, since `CommandFactories` is instantiated with `ensureTopicExistence == true`. But it would still be good to add a check for this.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java (hunk 1 above)
Comment (hjafarpour) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java]: Left over, removed it.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java (hunk 2 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: The other one was public for test purpose. I made this one private.

File: ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java (hunk 5 above)
Comment (hjafarpour) [reply ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java]: Added test to ensure we check the topic existence when it is needed.

**Comment by hjafarpour** [comment]
@dguy added a test for when we don't need to enforce topic existence.
@rodesai Updated the PR title.
@apurvam  Added test to ensure we enforce topic existence when it is needed.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 6]
```diff
@@ -238,7 +238,7 @@ public QueryMetadata getQueryExecutionPlan(final Query query) throws Exception {
   }
 
 
-  private List<Pair<String, Statement>> parseQueries(
+  public List<Pair<String, Statement>> parseQueries(
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: This can be package private. Should also add a `// Visible for Testing` comment.

File: ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java [hunk 7]
```diff
@@ -74,4 +79,49 @@ public void shouldFailWhenSyntaxIsInvalid() throws Exception {
...
+      List<QueryMetadata> queries =
+          ksqlEngine.buildMultipleQueries(ddlStatement.toString(), Collections.emptyMap());
+      Assert.fail();
+    } catch (Exception e) {
+      assertThat(e.getMessage(), equalTo("Kafka topic does not exist: S1_NOTEXIST"));
+    }
+  }
+
+  @Test
+  public void shouldEnforceTopicExistanceCorrectly() throws Exception {
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java]: Typo, `Existance` -> `Existence`. there are a few other places as well.
